Respond only to IPv4/ANY requests, others with NoData response.

oldium opened the pull request:
When the request with IPv6 (type `AAAA`) comes, it is responded as IPv4 (type `A`). Fix this by checking the request class type and allow only `ANY` or `A`. When there is no address, allow delaying the response for the case when more DNS servers have been asked to allow winning the first response with the actual address.

Before this PR:

* DNS server responded to all DNS query types, including AAAA (IPv6), MX, TXT, etc.
* Even in IPv4-only setups like Minikube, AAAA queries were answered with empty responses, leading to resolution failures, especially on Windows.
* No support for delaying NoData responses, causing issues when multiple DNS servers were queried in parallel.

After this PR:

* DNS server responds only to A (IPv4) and ANY queries; all others get a clean NoData response.
* Introduces support for dns-nodata-delay-ms via ConfigMap to delay empty responses and improve DNS resolution in multi-server setups.
* Responds with `SERVFAIL` DNS reply in case the name resolution fails.
* Adds live config reload with file watch and detailed logging for DNS requests, responses, and config updates.
* Upgrades base image to Node.js 24 (node:24-alpine).
* Migrates to stable Kubernetes RBAC APIs (rbac.authorization.k8s.io/v1).

oldium committed:
Respond only to IPv4/ANY requests, others with NoData response.

Signed-off-by: Oldřich Jedlička <oldium.pro@gmail.com>

sharifelgamal commented:
Hey @oldium, this looks like a reasonable PR. I'm gonna get someone to test this before we merge and release a new image.

oldium commented:
I am thinking about putting the A records into additional records field (instead of ignoring them) because this field is used for anticipated/related queries - it can be anticipated that the client asks for A query when AAAA does not exist.

oldium commented:
I will do it later this week, this can speed-up lookup.

oldium committed:
Allow delaying of NoData response in case multiple DNS servers are asked.

Signed-off-by: Oldřich Jedlička <oldium.pro@gmail.com>

oldium commented:
Updated the code a little bit to allow coexistence of multiple minikube instances and theirs DNS servers with overlapping host names.

In such overlapping case the OS needs to be configured to query all DNS servers (like `dnsmasq` is able to do) and the ingress-dns needs to configure DNS_NODATA_DELAY_MS value to add the delay before sending the actual NoData responses.

oldium commented:
Just updated logging, no functional changes.

oldium commented:
I tested it thoroughly as part of issue https://github.com/kubernetes/minikube/issues/14872 and this should be the final version, I think.

oldium commented:
It would be good to have the DNS_NODATA_DELAY_MS documented and have it configurable somehow - please suggest how.

oldium commented:
Updated to use ConfigMap and added some description. I hope this could be used somehow in minikube as well. If not, please suggest better solution. Thanks.

oldium commented:
Added file watcher, so that the new value from ConfigMap is automatically applied (this takes few seconds - the ConfigMap change is not visible immediately).

oldium commented:
Finished testing, ready to be used.

oldium commented:
Just small logging and formatting update, no functional changes. Done, from my point of view there is nothing more to polish at the moment :-)

oldium committed:
Added possibility to configure the NoData response delay.

The file watcher needs to be reinstalled when new configuration is loaded
(there is a new file with new value). For this it is good to abort the old
file watcher, which is possible since Node.js 15.9.0 and greater. Thus
update Node.js and while at it, update also dependencies.

Signed-off-by: Oldřich Jedlička <oldium.pro@gmail.com>

# Conflicts:
#	app/dns-server/nodejs/src/index.js

oldium commented:
Made some dependency updates, added `parseInt` fallbacks.

9SMTM6 commented:
@sharifelgamal is there something I can do to test this? I seem to have the same issue, and ended up having to permanently change priorization of ipv4 and ipv6.

oldium commented:
Roughly: You can build your own image (by building `./docker/nodejs/Dockerfile` from `/app/dns-server` directory, give it a tag name `minikube-ingress-dns`) and then apply it with the following Yaml to Minikube with `minikube kubectl -- apply`:

```yaml
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: minikube-ingress-dns
  namespace: kube-system
  labels:
    app: minikube-ingress-dns
    kubernetes.io/bootstrapping: rbac-defaults
    app.kubernetes.io/part-of: kube-system
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: minikube-ingress-dns
  namespace: kube-system
  labels:
    app: minikube-ingress-dns
    kubernetes.io/bootstrapping: rbac-defaults
    app.kubernetes.io/part-of: kube-system
    gcp-auth-skip-secret: "true"
rules:
  - apiGroups:
      - ""
      - "extensions"
      - "networking.k8s.io"
    resources:
      - ingresses
    verbs:
      - get
      - list
      - watch
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: minikube-ingress-dns
  namespace: kube-system
  labels:
    app: minikube-ingress-dns
    kubernetes.io/bootstrapping: rbac-defaults
    app.kubernetes.io/part-of: kube-system
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: minikube-ingress-dns
subjects:
  - kind: ServiceAccount
    name: minikube-ingress-dns
    namespace: kube-system
---
apiVersion: v1
kind: Pod
metadata:
  labels:
    app: minikube-ingress-dns
    app.kubernetes.io/part-of: kube-system
  name: kube-ingress-dns-minikube
  namespace: kube-system
spec:
  containers:
    - env:
        - name: DNS_PORT
          value: "53"
        - name: POD_IP
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: status.podIP
      image: minikube-ingress-dns
      imagePullPolicy: IfNotPresent
      name: minikube-ingress-dns
      ports:
        - containerPort: 53
          hostPort: 53
          protocol: UDP
      volumeMounts:
        - mountPath: /config
          name: minikube-ingress-dns-config-volume
  volumes:
    - name: minikube-ingress-dns-config-volume
      configMap:
        name: minikube-ingress-dns
  hostNetwork: true
  serviceAccountName: minikube-ingress-dns
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: minikube-ingress-dns
  namespace: kube-system
  labels:
    app: minikube-ingress-dns
    app.kubernetes.io/part-of: kube-system
data:
  dns-nodata-delay-ms: "20"
```

9SMTM6 commented:
Sorry for taking a bit with my answer, largely personal stuff.

It seems to work, after a bit of trial and error.

First I removed the registry edit I did to change the preference to ipv4 over ipv6 and rebooted.

I did `minikube addons disable ingress-dns`, then I build the docker file in the minikube node (my windows host doesnt have a docker cli): `minikube ssh`, git clone of your repo (https://github.com/oldium/minikube-ingress-dns.git), `cd minikube-ingress-dns/app/dns-server/`, then I built the image as instructed (`docker build -t minikube-ingress-dns -f docker/nodejs/Dockerfile .`, and then I exited the minikube node and applied the provided config.

After a bit of waiting I could reach the NS, and as intended it provided a response to A-record requests and none to AAAA: With `kubectl apply -f https://raw.githubusercontent.com/kubernetes/minikube/master/deploy/addons/ingress-dns/example/example.yaml` I tested the dns server directly first: `nslookup -type=AAAA hello-john.test $(minikube ip)` yields refusal, `nslookup -type=A hello-john.test $(minikube ip)` answers with the expected.

Then the final test, that is the important one: After applying the powershell line [here](https://minikube.sigs.k8s.io/docs/handbook/addons/ingress-dns/#installation) I could open `http://hello-john.test` in my browser and got the expected answer.

@sharifelgamal is there something else I should test? @oldium seems to have put some effort into this, and I can attest that this issue is not unseen. And, to my understanding, this isn't a fix for some idiosyncracy of Windows, but instead this brings ingress-dns to conformity with the standards. I beliefe @oldium would be thankful to be able to stop bumping the issue in https://github.com/kubernetes/minikube/issues/14872 to keep it from being marked stale.

oldium commented:
@9SMTM6 Thanks for testing and for the How-To, really nice job!

adminnz commented:
I'm having the same issue. I've tried the what @9SMTM6 said in terms of building the repo and deploying.
But I cannot get DNS to return any results.

`nslookup -type=A hello-john $(minikube ip)`

Just returns
```
DNS request timed out.
    timeout was 2 seconds.
Server:  UnKnown
Address:  192.168.49.2

DNS request timed out.
    timeout was 2 seconds.
DNS request timed out.
    timeout was 2 seconds.
*** Request to UnKnown timed-out
```

My minikube ip is 192.168.49.2

`kubectl get pods -n kube-system`
returns
```
NAME                               READY   STATUS    RESTARTS      AGE
coredns-7db6d8ff4d-w9hw7           1/1     Running   1 (23m ago)   20h
etcd-minikube                      1/1     Running   1 (23m ago)   20h
kube-apiserver-minikube            1/1     Running   1 (23m ago)   20h
kube-controller-manager-minikube   1/1     Running   1 (23m ago)   20h
kube-ingress-dns-minikube          1/1     Running   0             10m
kube-proxy-xdz9j                   1/1     Running   1 (23m ago)   20h
kube-scheduler-minikube            1/1     Running   1 (23m ago)   20h
metrics-server-c59844bb4-h45pv     1/1     Running   0             23m
storage-provisioner                1/1     Running   3 (23m ago)   20h
```


Pod logs:
```
kubectl logs kube-ingress-dns-minikube -n kube-system
yarn run v1.22.19
$ node src/index.js
Watching config file /config/dns-nodata-delay-ms for changes
Listening to 192.168.49.2 on port 53 and NoData packet delay 20 ms
```

9SMTM6 commented:
The windows setup I used doesn't exist any longer, so I can not test that stuff.

But your nslookup doesnt have the `.test` top-level-domain. If that wasn't a mistake while explaining what you did that could certainly be the simple issue.

adminnz commented:
That was just a typo in the post.
I've done all the steps as far as I can tell, and nothing works (nslookup / test-netconnection / ping / browser)

```
PS C:\Users\admin> Get-DnsClientNrptRule | Where-Object {$_.Namespace -eq '.test'} | Remove-DnsClientNrptRule -Force; Add-DnsClientNrptRule -Namespace ".test" -NameServers "$(minikube ip)"
PS C:\Users\admin> nslookup -type=AAAA hello-john.test $(minikube ip)
DNS request timed out.
    timeout was 2 seconds.
Server:  UnKnown
Address:  192.168.49.2

DNS request timed out.
    timeout was 2 seconds.
DNS request timed out.
    timeout was 2 seconds.
*** Request to UnKnown timed-out
PS C:\Users\admin> nslookup -type=A hello-john.test $(minikube ip)
DNS request timed out.
    timeout was 2 seconds.
Server:  UnKnown
Address:  192.168.49.2

DNS request timed out.
    timeout was 2 seconds.
DNS request timed out.
    timeout was 2 seconds.
*** Request to UnKnown timed-out
PS C:\Users\admin> Get-DnsClientNrptRule

Name                             : {A8DFAB4E-7BEA-4665-A018-51CA9C81BB09}
Version                          : 2
Namespace                        : {.test}
IPsecCARestriction               :
DirectAccessDnsServers           :
DirectAccessEnabled              : False
DirectAccessProxyType            :
DirectAccessProxyName            :
DirectAccessQueryIPsecEncryption :
DirectAccessQueryIPsecRequired   :
NameServers                      : 192.168.49.2
DnsSecEnabled                    : False
DnsSecQueryIPsecEncryption       :
DnsSecQueryIPsecRequired         :
DnsSecValidationRequired         :
NameEncoding                     : Disable
DisplayName                      :
Comment                          :

PS C:\Users\admin>
```

Its like anything in the minikube cannot be reached from my computer.
Is the minikube ingress IP address meant to be on the same network as my computer because my computer is on a network with 192.168.1.1/24

oldium commented:
I have checked my minikube auto-startup script and you also need to enable ingress with `minikube.exe addons enable ingress`

9SMTM6 commented:
I'm no network expert. But to be able to do a DNS request, or any request, you of course need to be able to reach the server.

This setup was tested with the hyperv driver. Should PROBABLY also work with Docker and podman.

These tests were started from a state where the minikube setup was working when I switched my windows setup to use preferably ipv4. Thus I had already enabled ingress in minikube. I should've clarified that, I'll add it as edit to the original steps @oldium.

adminnz commented:
I'm using the docker driver, but the same thing happens with the hyperv driver, and yes i have the ingress addon enabled.
I don't understand why following the minikube tutorial doesn't work.
Its also not just me, I've had a colleague try and get it working and he also gets the same result (fresh install of docker/minikube)

9SMTM6 commented:
Did you try the new step 0 in my testing steps? Afterwards it should work.

This PR is primarily because 1.this fix is suboptimal and 2. The steps in the official guide don't work with modern windows.

oldium commented:
I don't recall taking any action concerning IPv6. I'll give it a try this week; it's been a long time since I last used minikube. The key point here is to overcome the timeout and get the real DNS response.

oldium commented:
Confirmed, it works. It looks like you simply skipped some steps. All are mandatory 😅

I replayed what @9SMTM6 wrote almost one-to-one, check the following steps, some steps require administrator PowerShell. You need to create file `ingress-dns.yaml` with the content of https://github.com/sharifelgamal/minikube-ingress-dns/pull/4#issuecomment-1919208524.
No kubectl nor docker is necessary, just have `minikube.exe` installed.

I started with clean minikube environment and clean DNS rules just to be sure (first two commands):

```
#> minikube.exe delete --purge

# Run as admin - remove all DNS rules
#> Get-DnsClientNrptRule | Remove-DnsClientNrptRule

#> minikube.exe addons enable ingress

# Check enabled addons
#> (minikube.exe addons list -o json | ConvertFrom-Json).PSObject.Properties | ? { $_.Value.Status -eq 'enabled' } | % { $_.Name }
default-storageclass
ingress
storage-provisioner

#> minikube.exe ssh
git clone https://github.com/oldium/minikube-ingress-dns.git
cd minikube-ingress-dns/app/dns-server
docker build -t minikube-ingress-dns -f docker/nodejs/Dockerfile .
exit

#> minikube.exe kubectl -- apply -f .\ingress-dns.yaml
serviceaccount/minikube-ingress-dns created
clusterrole.rbac.authorization.k8s.io/minikube-ingress-dns created
clusterrolebinding.rbac.authorization.k8s.io/minikube-ingress-dns created
pod/kube-ingress-dns-minikube created
configmap/minikube-ingress-dns created

#> minikube.exe kubectl -- apply -f https://raw.githubusercontent.com/kubernetes/minikube/master/deploy/addons/ingress-dns/example/example.yaml
deployment.apps/hello-world-app created
ingress.networking.k8s.io/example-ingress created
service/hello-world-app created
service/hello-world-app created

#> nslookup -type=AAAA hello-john.test (minikube.exe ip)
Server:  UnKnown
Address:  172.30.7.17

*** No IPv6 address (AAAA) records available for hello-john.test

#> nslookup -type=A hello-john.test (minikube.exe ip)
Server:  UnKnown
Address:  172.30.7.17

Non-authoritative answer:
Name:    hello-john.test
Address:  172.30.7.17

# Run as admin
#> Add-DnsClientNrptRule -Namespace '.test' -NameServers (minikube.exe ip)

#> Get-DnsClientNrptRule                      

Name                             : {95FDC8E0-3887-4AA6-884B-718152962519}
Version                          : 2
Namespace                        : {.test}
IPsecCARestriction               :
DirectAccessDnsServers           :
DirectAccessEnabled              : False
DirectAccessProxyType            :
DirectAccessProxyName            :
DirectAccessQueryIPsecEncryption :
DirectAccessQueryIPsecRequired   :
NameServers                      : 172.30.7.17
DnsSecEnabled                    : False
DnsSecQueryIPsecEncryption       :
DnsSecQueryIPsecRequired         :
DnsSecValidationRequired         :
NameEncoding                     : Disable
DisplayName                      :
Comment                          :

#> ping hello-john.test

Pinging hello-john.test [172.30.7.17] with 32 bytes of data:
Reply from 172.30.7.17: bytes=32 time<1ms TTL=64
```

oldium commented:
Finished fine-tuning the previous post 😊. It should have all necessary info. Feel free to ask in case of doubt.

adminnz commented:
@oldium OMG Thank you so much. :sparkling_heart: It is working now!

sharifelgamal commented:
hey, I haven't worked on minikube in quite some time. I'll get @medyagh to merge and release.

medyagh commented:
hi @oldium @kartikjoshi21 can you post int he PR description the Before/After this PR ?

kartikjoshi21 commented:
> hi @oldium @kartikjoshi21 can you post int he PR description the Before/After this PR ?

@medyagh i dont have permission to update this PR. I think we can include this in PR desciption regarding before/After. Also i tested this PR and im able to resolve dns queries perfectly on windows + Minikube + hyperv setup. 

**Before this PR:**

- DNS server responded to all DNS query types, including AAAA (IPv6), MX, TXT, etc.
- Even in IPv4-only setups like Minikube, AAAA queries were answered with empty responses, leading to resolution failures, especially on Windows.
- No support for delaying NoData responses, causing issues when multiple DNS servers were queried in parallel.

**After this PR:**

- DNS server responds only to A (IPv4) and ANY queries; all others get a clean NoData response.
- Introduces support for dns-nodata-delay-ms via ConfigMap to delay empty responses and improve DNS resolution in multi-server setups.
- Adds live config reload with file watch and detailed logging for DNS requests, responses, and config updates.
- Upgrades base image to Node.js 16 (node:16-alpine).
- Migrates to stable Kubernetes RBAC APIs (rbac.authorization.k8s.io/v1).

oldium committed:
Migration to latest dependencies

Updates:

* Updated to Node.js 24 (next LTS)
* Updated `nodemon` and `@kubernetes/client-node` to latest versions
* Migrated from deprecated `request` library to `node-fetch` library
* Migrated code to work with latest `@kubernetes/client-node`
* When DNS resolution fails, send SERVFAIL error DNS response

Signed-off-by: Oldřich Jedlička <oldium.pro@gmail.com>

oldium commented:
Updated to latest dependencies:

* Updated to Node.js 24 (next LTS)
* Updated `nodemon` and `@kubernetes/client-node` to latest versions
* Removed `request` dependency
* Migrated code to work with latest `@kubernetes/client-node`
* Migrated to Fetch API to remove obsolete dependency on `request` library
* When DNS resolution fails, send SERVFAIL error DNS response

Tested only locally currently (in WebStorm IDE) without a cluster, I need to test it with minikube.

oldium commented:
Hm. Fetching fails with `minikube.exe` due to certificates (`unable to verify the first certificate; if the root CA is installed locally, try running Node.js with --use-system-ca`), so the HTTP client setup is wrong. I will check it.

oldium commented:
Fixed. I migrated to the `node-fetch` API instead of the native Node.js Fetch API. The native Fetch API is more limited in terms of certificates, hence `node-fetch`. Also, the K8s JavaScript client seems to expect [`node-fetch` library](https://github.com/kubernetes-client/javascript/blob/main/examples/raw-example.js#L2).

kartikjoshi21 commented:
Thanks @oldium , @medyagh can we plan to merge this now ?
